mqtt/mqtt.js

```diff
@@ -87,125 +57,77 @@ nats.subscribe('channel.*', function (msg) {
-               */
-              Publisher: client.id,
-              Channel: channel,
-              Protocol: 'mqtt',
-              ContentType: packet.topic.split('/')[4],
-              Payload: packet.payload
+    // Topics are in the form `channels/<channel_id>/messages/senml-json`
+    var channel = packet.topic.split('/')[1];
+
+    things.CanAccess({
+        token: client.password,
+        chanID: channel
+    }, function (err, res) {
+        if (!err) {
+            logger.info('authorized publish');
+            /**
+             * We must publish on NATS here, because on_publish() is also called
+             * when we receive message from NATS from other adapters (in nats.subscribe()),
+             * so we must avoid re-publishing on NATS what came from other adapters
+             */
+            var rawMsg = message.RawMessage.encode({
+                Publisher: client.id,
+                Channel: channel,
+                Protocol: 'mqtt',
+                ContentType: packet.topic.split('/')[3],
+                Payload: packet.payload
             });
 
-            console.log(msg);
-
-            console.log(util.inspect(packet, false, null))
-
             // Pub on NATS
-            nats.publish('channel.' + channel, msg);
+            nats.publish('channel.' + channel, rawMsg);
         } else {
-            console.log('Publish not authorized');
-            error = 4; // Bad username or password
+            logger.warn("unauthorized publish: %s", err.message);
+            callback(4); // Bad username or password
         }
-        callback(error);
     });
 };
 
 // AuthZ SUB
 aedes.authorizeSubscribe = function (client, packet, callback) {
-    // Topics are in the form `mainflux/channels/<channel_id>/messages/senml-json`
-    var channel = packet.topic.split('/')[2];
-    /**
-    * Check if PUB is authorized
-    */
-    var options = {
-        url: config.auth_url + ':' + config.auth_port + '/channels/' + channel + '/access-grant',
-        method: 'GET',
-        headers: {
-            'Content-Type': 'application/json',
-            'Authorization': client.password
-        }
-    };
-
-    request(options, function (err, res) {
-        var error = null;
-        if (res && (res.statusCode === 200)) {
-            console.log('Subscribe authorized OK');
+    // Topics are in the form `channels/<channel_id>/messages/senml-json`
+    var channel = packet.topic.split('/')[1];
+    
+    things.canAccess({
+        token: client.password,
+        chanID: channel
+    }, function (err, res) {
+        if (!err) {
+            logger.info('authorized subscribe');
+            callback(null, packet);
         } else {
-            console.log('Subscribe not authorized');
-            error = 4; // Bad username or password
+            logger.warn('unauthorizerd subscribe: %s', err);
+            callback(4, packet); // Bad username or password
```

Comment:
Same here for callback.

Reply:
Here, `callback` is called if there is no error. Check `if` part.